src/goto-symex/symex_main.cpp

```diff
@@ -67,17 +67,65 @@ void symex_transition(
     // 1. the transition from state.source.pc to "to" is not a backwards goto
     // or
     // 2. we are arriving from an outer loop
+
+    // TODO: This should all be replaced by natural loop analysis.
     framet &frame = state.call_stack().top();
     const goto_programt::instructiont &instruction=*to;
     for(const auto &i_e : instruction.incoming_edges)
+    {
       if(i_e->is_goto() && i_e->is_backwards_goto() &&
          (!is_backwards_goto ||
           state.source.pc->location_number>i_e->location_number))
       {
         const auto loop_id =
           goto_programt::loop_id(state.source.function_id, *i_e);
-        frame.loop_iterations[loop_id].count = 0;
+        auto &current_loop_info = frame.loop_iterations[loop_id];
+        current_loop_info.count = 0;
+        current_loop_info.is_active = true;
+
+        // Work out what parent this loop has.
+        for(auto &loop : frame.loop_iterations)
+        {
+          if(
+            loop.second.is_active && !loop.second.is_recursion &&
+            current_loop_info.nesting_level < loop.second.nesting_level)
```

Comment:
This reads backwards -- I'd expect this to mean "current_loop_info's parent is now loop.second", but this is surely the other way around given `loop.second` is *more* deeply nested

Reply:
Also if this loop was nested at depth 3 this would match both our parent and our grandparent loop, right? It would pick one arbitrarily?

Reply:
Overall this looks a lot like we should be maintaining an active loop stack rather than looking at all currently active loops and their depths.

Reply:
In regards to maintaining an active loop stack: its aim was to do that without an explicit stack we have to modify and keep track of, though at this point it might be worth just doing it, as it's grown more complicated than I intended (and may get more-so). 

Also as you noticed, small issue here that it should be <= in `current_loop_info.nesting_level < loop.second.nesting_level`. since if we've found a equal nesting level we want to keep walking upwards.